s3path.py

```diff
@@ -205,23 +180,57 @@ def listdir(self, path):
             return [entry.name for entry in scandir_iter]
 
     def open(self, path, *, mode='r', buffering=-1, encoding=None, errors=None, newline=None):
-        bucket_name = self.bucket_name(path.bucket)
-        key_name = str(path.key)
-        resource, _ = self.configuration_map.get_configuration(path)
-        object_summery = resource.ObjectSummary(bucket_name, key_name)
-        file_object = S3KeyReadableFileObject if 'r' in mode else S3KeyWritableFileObject
-        return file_object(
-            object_summery,
-            path=path,
+        resource, config = self.configuration_map.get_configuration(path)
+
+        def get_resource_kwargs():
+            # This is a good example of the complicity of boto3 and botocore
+            # resource arguments from the resource object :-/
+            # very annoying...
+
+            try:
+                access_key = resource.meta.client._request_signer._credentials.access_key
+                secret_key = resource.meta.client._request_signer._credentials.secret_key
+                token = resource.meta.client._request_signer._credentials.token
+            except AttributeError:
+                access_key = secret_key = token = None
+            return {
+                'endpoint_url': resource.meta.client.meta._endpoint_url,
+                'config': resource.meta.client._client_config,
+                'region_name': resource.meta.client._client_config.region_name,
+                'use_ssl': resource.meta.client._endpoint.host.startswith('https'),
+                'verify': resource.meta.client._endpoint.http_session._verify,
+                'aws_access_key_id': access_key,
+                'aws_secret_access_key': secret_key,
+                'aws_session_token': token,
+            }
+
+        dummy_object = self._s3.Object('bucket', 'key')
+        object_kwargs = self._update_kwargs_with_config(
+            dummy_object.get, config=config)
+        multipart_upload_kwargs = self._update_kwargs_with_config(
+            dummy_object.initiate_multipart_upload, config=config)
+
+        file_object = smart_open.open(
+            uri=path.as_uri(),
             mode=mode,
             buffering=buffering,
             encoding=encoding,
             errors=errors,
-            newline=newline)
+            newline=newline,
+            ignore_ext=True,
+            transport_params={
+                'session': boto3.DEFAULT_SESSION,
+                'resource_kwargs': get_resource_kwargs(),
```

Comment:
```python
        s3path_session = boto3.Session()
        s3path_client = s3path_session.client(
            's3',
            endpoint_url=resource_kwargs['endpoint_url'],
            aws_access_key_id=resource_kwargs['aws_access_key_id'],
            aws_secret_access_key=resource_kwargs['aws_secret_access_key'],
            aws_session_token=resource_kwargs['aws_session_token']
        )
        file_object = smart_open.open(
            uri=path.as_uri(),
            mode=mode,
            buffering=buffering,
            encoding=encoding,
            errors=errors,
            newline=newline,
            ignore_ext=True,
            transport_params={
                'session': s3path_session,
                'client': s3path_client,
                'resource_kwargs': resource_kwargs,
```

resource_kwargs seems not working when using custom endpoint , I set it on transport_params['client'] and transport_params['session'] instead and it worked ! 

please consider my method , this is also on offical readme page.

Reply:
Hi @LeoQuote and @cooperlees sorry for the delay in response.

Great catch!
Let's add your fix as well.

Regarding timeline, I'll start writing the changelog for this big version.
After that you will have a new version. :-)

ETA - two days +-

Reply:
Awesome. Many thanks. Let’s us know if we can help. We’ll be testing with bandersnatch soon as released.